k8s/upgrade/src/plugin/upgrade.rs

```diff
@@ -103,25 +103,66 @@ pub struct UpgradeArgs {
     /// Upgrade to an unsupported version.
     #[clap(global = true, hide = true, long, default_value_t = false)]
     pub skip_upgrade_path_validation_for_unsupported_version: bool,
+
+    /// Upgrade to an unsupported version.
+    #[clap(global = true, long)]
+    pub set: Vec<String>,
+}
+
+impl Default for UpgradeArgs {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
 impl UpgradeArgs {
+    /// Initialise with default values.
+    pub fn new() -> Self {
+        Self {
+            dry_run: false,
+            skip_data_plane_restart: false,
+            skip_single_replica_volume_validation: false,
+            skip_replica_rebuild: false,
+            skip_cordoned_node_validation: false,
+            skip_upgrade_path_validation_for_unsupported_version: false,
+            set: Default::default(),
+        }
+    }
     ///  Upgrade the resources.
-    pub async fn apply(&self, namespace: &str) {
+    pub async fn apply(&self, namespace: &str) -> error::Result<()> {
+        let gur = UpgradeEventClient::new(namespace).await?;
+        let release_name = get_release_name(namespace).await?;
+
+        // Delete any previous upgrade events
+        gur.delete_upgrade_events(release_name.clone()).await?;
         // Create resources for upgrade
-        _ = UpgradeResources::create_upgrade_resources(
-            namespace,
-            self.skip_data_plane_restart,
-            self.skip_upgrade_path_validation_for_unsupported_version,
-        )
-        .await
-        .map_err(|error| {
-            std::process::exit(error.into());
-        });
-        console_logger::info(UPGRADE_JOB_STARTED, "");
+        UpgradeResources::create_upgrade_resources(namespace, self).await?;
+
+        for _i in 0 .. 6 {
+            tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+            match gur.get_upgrade_events(release_name.clone(), false).await {
+                Ok(latest_event) => {
+                    if latest_event.type_ == Some("Warning".to_string()) {
```

Comment:
Rather than checking for first error being a warning, should we have some specific details here about validation failed?

Reply:
```
 // Delete any previous upgrade events
 gur.delete_upgrade_events(release_name.clone()).await?;
```
The above code actually deletes any upgrade related events before creating the job , so there will not be any upgrade related events before teh job starts. And on the job side the event is published as action as Error or others. So made the code change where matching id cone of actions for Error. 
https://github.com/openebs/mayastor-extensions/blob/5c35c902f9d01df7bb7593334ff5bcacd5563ebf/k8s/upgrade/src/bin/upgrade-job/events/event_recorder.rs#L259